### PR TITLE
Add new Relax function to the batched model for evaluating query tokens over multiple time steps in parallel 

### PR DESCRIPTION
**Pull request by masahi**
In speculative decoding and restoring KV cache entries for evicted parallel-sampling requests, we need to be able to compute logits over multiple tokens (time steps) while utilizing the KV cache for the past tensors. This is a hybrid of `prefill` and `decode` functions, in that
* `prefill` can compute logits over multiple tokens but doesn't read from KV cache
* `decode` works on one token at a time.  

I'm introducing a new function, tentatively called `evaluate_multi_query`, for this purpose. `multi_query_decode` is also a good name.

The changes in `run_llama_batched_vllm.py` shows a new request type and how the new function is meant to be used. There is no change under `serve` yet since it is purely a model change. After we agree on the approach, I'll integrate this new function into the engine to complete my parallel-sampling work. @yelite needs this for speculative decoding.

There is no attention kernel that reads from KV cache and operates on multiple queries, except FlashInfer which has `BatchedPrefillWithKVCache`. But we can emulate the behavior of such kernel by materializing past KV tensors from the cache, concat them with the present tensors, and running the standard prefill attention. This is not efficient but its correctness is much easier to verify. Until we integrate FlashInfer or Flash attention adds paged KV cache support, we can use this emulation.

@sunggg @yelite @elvin-n

### CODE REVIEW

File: examples/python/run_llama_batched_vllm.py [hunk 1]
```diff
@@ -241,6 +248,71 @@ def _pad_to_max(x: List[int], max_len: int) -> List[int]:
...
+        query_lens.append(num_queries)
+        request_id = request.request_id
+        input_ids += request.query_token_ids
+
+        positions += [num_past_tokens + i for i in range(num_queries)]
+
+        if sliding_window:
+            seq_lens.append(min(num_past_tokens + num_queries, sliding_window))
+            num_past = min(num_past_tokens, sliding_window)
+            past_slot_mapping += all_slot_mappings[request_id][num_past:]
```
Comment (yelite) [inline examples/python/run_llama_batched_vllm.py]: Should this be `[:num_past]`?

Comment (masahi) [reply examples/python/run_llama_batched_vllm.py]: yes fixed now

I haven't tested the sliding-window code path yet. Things get complicated especially with parallel sampling + SWA case.

**Comment by sunggg** [comment]
Thank you for the PR, @masahi! Which tvm should I use to run this? 
Also, would you move `examples/python/run_llama_batched_vllm.py` under the `serve/` so that that can be a single folder for mlc-serve?

**Comment by masahi** [comment]
For now we need TVM from https://github.com/masahi/tvm/tree/vllm-cache-reconstruct. After https://github.com/apache/tvm/pull/16376 is merged, I'll do a rebase. 

>  would you move examples/python/run_llama_batched_vllm.py under the serve/ so that that can be a single folder for mlc-serve?

`examples/python/run_llama_batched_vllm.py` is not associated with `mlc-serve`. `mlc-ai/main` also has it. I added it to demonstrate how to use the batched llama model.

File: examples/python/run_llama_batched_vllm.py (hunk 1 above)
Comment (masahi) [reply examples/python/run_llama_batched_vllm.py]: Actually above is still not correct when `num_past_tokens + num_queries > sliding_window`. I believe the correct one is
```
if sliding_window and num_past_tokens + num_queries >= sliding_window:
    seq_lens.append(sliding_window)
    past_slot_mapping += all_slot_mappings[request_id][
        num_past_tokens - (sliding_window - num_queries) : num_past_tokens
    ]
```

@yelite Can you check that this looks ok?

**Comment by masahi** [comment]
Opened https://github.com/octoml/mlc-llm/pull/157 which uses the new Relax function from this PR to enable parallel-sampling eviction.

File: examples/python/run_llama_batched_vllm.py (hunk 1 above)
Comment (yelite) [reply examples/python/run_llama_batched_vllm.py]: Yes I think this one is correct. `past_slot_mapping` has `sliding_window - num_queries` in it and `slot_mapping` has the remaining `num_queries` tokens

**Review by yelite — APPROVED** [review]
This looks great and should be sufficient for the speculative decoding with draft model. 

By the way, is it still necessary to keep `decode` after we have a good kernel for `evaluate_multi_query`? Will there be performance loss if we run `evaluate_multi_query` with one token from each sequence? If not, maybe we can just name this `decode`. Maybe we can even retire `prefill` if the kernel can specialize without degrading performance in the case where it doesn't need to read past KV from cache.

**Comment by masahi** [comment]
> By the way, is it still necessary to keep decode after we have a good kernel for evaluate_multi_query? Will there be performance loss if we run evaluate_multi_query with one token from each sequence? If not, maybe we can just name this decode. Maybe we can even retire prefill if the kernel can specialize without degrading performance in the case where it doesn't need to read past KV from cache.

This is an interesting idea. I'd like to think that specialization allows perf advantages (`decode` kernel shouldn't parallelize over the query tokens, since that dimension is small). FlashInfer implements a dedicated kernel for batched decode while it also has `BatchedPrefillWithKVCache`.  We have to measure and see. 

The comparison is a bit subtle since moving from single query to multiple ones involves switching entirely different kernel implementations (vllm to flash attention / flash infer). So perf can be affected by any number of reasons besides the increase in the number of query tokens.
